sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -92,18 +114,20 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
           // Deduplicate conflicting attributes if any.
           dedupJoin(Join(outerPlan, sub, LeftAnti, joinCond))
```

Comment:
Looks like we don't need `dedupJoin`, but always dedup the subquery before putting it in a join.

Reply:
I think it makes sense to dedup the subquery only when the join condition has not been created yet (so in the case of `InSubquery`). In this case, the condition is already there, so I think we still have to use `dedupJoin` (for `Exists`)

Reply:
`dedupJoin` will evetually dedup the subquery IIUC.

What I'd like to do is to unify `dedupJoin` and `dedupSubqueryOnSelfJoin`, so that the code will be consistent for all cases:
```
val newSub = dedup(sub, values)
// create join condition if any
Join(outerPlan, newSub, ...)
```

Reply:
the main problem is that in the other cases, so when exists is there, the condition is already created. So we would need to complicate quite a lot the method in order to handle the 2 cases and I am not sure wether it is worth. For instance, the `values`, in the `Exists` case, should be taken from the conditions as those expressions referencing attributes from one side and the join condition needs to be rewritten. So I don't think that it is a good idea to have a common rewrite for both them: it would be overcomplicated IMHO.

Reply:
ah thanks for the explanation! Makes sense to me.